engineapi/backing_image_manager.go

```diff
@@ -0,0 +1,175 @@
+package engineapi
+
+import (
+	"fmt"
+	bimapi "github.com/longhorn/backing-image-manager/api"
+	bimclient "github.com/longhorn/backing-image-manager/pkg/client"
+	"github.com/longhorn/longhorn-manager/types"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+)
+
+const (
+	BackingImageManagerDefaultPort = 8000
+
+	CurrentBackingImageManagerAPIVersion = 1
+	UnknownBackingImageManagerAPIVersion = 0
+)
+
+type BackingImageManagerClient struct {
+	ip            string
+	apiMinVersion int
+	apiVersion    int
+
+	grpcClient *bimclient.BackingImageManagerClient
+}
+
+func CheckBackingImageManagerCompatibilty(bimMinVersion, bimVersion int) error {
+	if CurrentBackingImageManagerAPIVersion > bimVersion || CurrentBackingImageManagerAPIVersion < bimMinVersion {
+		return fmt.Errorf("current BackingImageManager version %v is not compatible with BackingImageManagerAPIVersion %v and BackingImageManagerAPIMinVersion %v",
+			CurrentBackingImageManagerAPIVersion, bimVersion, bimMinVersion)
+	}
+	return nil
+}
+
+func NewBackingImageManagerClient(bim *longhorn.BackingImageManager) (*BackingImageManagerClient, error) {
+	if bim.Status.CurrentState != types.BackingImageManagerStateRunning || bim.Status.IP == "" {
+		return nil, fmt.Errorf("invalid Backing Image Manager %v, state: %v, IP: %v", bim.Name, bim.Status.CurrentState, bim.Status.IP)
+	}
+	if bim.Status.APIMinVersion != UnknownBackingImageManagerAPIVersion {
```

Comment:
Is it possible that the `bim.Status.CurrentState == types.BackingImageManagerStateRunning` and the `bim.Status.APIMinVersion != UnknownBackingImageManagerAPIVersion` ?  I see that we did a check here https://github.com/longhorn/longhorn-manager/blob/08c89c2d5e36f5585df7939bbbdb94fe47562f45/controller/backing_image_manager_controller.go#L459

Reply:
For the 1st time a bim becomes state running, `bim.Status.APIMinVersion` is `UnknownBackingImageManagerAPIVersion` and the controller needs to launch this gRPC client to get the actual version.

Once `bim.Status.APIMinVersion` is updated, this check will be applied each time before the controller wants to talk with the gRPC server.

Reply:
Thanks! My bad